src/completion.zsh

```diff
@@ -0,0 +1,156 @@
+      if [[ "${line[1]}" == "." ]]; then
+        continue
+      fi
+      echo "$line"
+    done
+  else
+    dir=$(dirname -- "$1")
+    length=$(echo -n "$dir" | wc -c)
+    if [ "$dir" = "/" ]; then
+      length=0
+    fi
+    seg=$(basename -- "$1")
+    starts_with_dir=$( \
+      find -L "$dir" -mindepth 1 -maxdepth 1 -type d \
+          2>/dev/null | cut -b $(( ${length} + 2 ))- | \sed '/^$/d' \
+          | while read -r line; do
+        if [[ "${seg[1]}" != "." && "${line[1]}" == "." ]]; then
+          continue
+        fi
+        if [[ "$line" == "$seg"* ]]; then
+          echo "$line"
+        fi
+      done
+    )
+    if [ -n "$starts_with_dir" ]; then
+      echo "$starts_with_dir"
+    else
+      find -L "$dir" -mindepth 1 -maxdepth 1 -type d \
+          2>/dev/null | cut -b $(( ${length} + 2 ))- | \sed '/^$/d' \
+          | while read -r line; do
+        if [[ "${seg[1]}" != "." && "${line[1]}" == "." ]]; then
+          continue
+        fi
+        if [[ "$line" == *"$seg"* ]]; then
+          echo "$line"
+        fi
+      done
+    fi
+  fi
+}
+
+__enhancd::completion::completer() {
+  case $ENHANCD_COMPLETION_BEHAVIOR in
+    ("list" | "dir" | "dirlist")
+        __enhancd::completion::list "${(Q)@[-1]}" | sort
+        ;;
+    ("history")
+        __enhancd::history::list "${(Q)@[-1]}"
+        ;;
+    ("default" | "*")
+        :
+        ;;
+  esac
+}
+
+__enhancd::completion::complete() {
+  setopt localoptions nonomatch
+  local l matches filter tokens base
+
+  l=$(__enhancd::completion::completer $@)
+
+  if [ -z "$l" ]; then
+    zle ${ENHANCD_COMPLETION_DEFAULT:-expand-or-complete}
+    return
+  fi
+
+  filter=$(__enhancd::utils::filter "$ENHANCD_FILTER")
+
+  if [ $(echo $l | wc -l) -eq 1 ]; then
+    matches=${(q)l}
+  else
+    matches=$(echo $l \
+        | FZF_DEFAULT_OPTS="--height ${FZF_TMUX_HEIGHT:-40%} \
+          --reverse $FZF_DEFAULT_OPTS $FZF_COMPLETION_OPTS \
+          --bind 'shift-tab:up,tab:down'" ${=filter} \
+        | while read -r item; do
+      echo -n "${(q)item} "
+    done)
+  fi
+
+  matches=${matches% }
+  if [ -n "$matches" ]; then
+    tokens=(${(z)LBUFFER})
+    base="${(Q)@[-1]}"
+    if [[ "$base" != */ ]]; then
+      if [[ "$base" == */* ]]; then
+        base="$(dirname -- "$base")"
+        if [[ ${base[-1]} != / ]]; then
+          base="$base/"
+        fi
+      else
+        base=""
+      fi
+    fi
+    LBUFFER="${tokens[1]} "
+    if [ -n "$base" ]; then
+      base="${(q)base}"
+      if [ "${tokens[2][1]}" = "~" ]; then
+        base="${base/#$HOME/~}"
+      fi
+      LBUFFER="${LBUFFER}${base}"
+    fi
+    LBUFFER="${LBUFFER}${matches}/"
+  fi
+  zle redisplay
+  typeset -f zle-line-init >/dev/null && zle zle-line-init
+}
+
+__enhancd::completion::run() {
+  setopt localoptions noshwordsplit noksh_arrays noposixbuiltins
+  local tokens cmd
+
+  tokens=(${(z)LBUFFER})
+  cmd=${tokens[1]}
+
+  if [[ "$LBUFFER" =~ "^\ *cd$" ]]; then
```

Comment:
Awesome work, but this regular expression is too restrictive. In particular, it should not check for `cd` but `$ ENHANCD_COMMAND`.

Reply:
Nice catch!! I'll update those 👍 @Konfekt

Reply:
Done 7f6dcf0973b5d448ecb8adebda96f28024965249